bin/picom-trans

```diff
@@ -46,194 +54,306 @@ print_usage() {
-  lineno=$(echo -n "$treeout" | grep -nw "^\s*$wid" | head -n1 | cut -d ':' -f 1)
+  # If it's already the topmost window
+  if echo "$treeout" | grep -q 'Parent window id: 0x[[:xdigit:]]* (the root window)'; then
+    topmost=$wid
+  else
+    # Get the whole window tree
+    treeout=$(xwininfo -root -tree)
+
+    if test -z "$treeout"; then
+      echo 'Failed to get root window tree.'
+      exit 1
+    fi
+
+    # Find the line number of the target window in the window tree
+    lineno=$(echo -n "$treeout" | grep -nw "^\s*$wid" | head -n1 | cut -d ':' -f 1)
 
-  if test -z "$lineno"; then
-    echo 'Failed to find window in window tree.'
+    if test -z "$lineno"; then
+      echo 'Failed to find window in window tree.'
+      exit 1
+    fi
+
+    # Find the highest ancestor of the target window below
+    topmost=$(echo -n "$treeout" \
+      | head -n $lineno \
+      | sed -n 's/^     \(0x[[:xdigit:]]*\).*/\1/p' \
+      | tail -n 1)
+  fi
+
+  if test -z "$topmost"; then
+    echo 'Failed to find the highest parent window below root of the' \
+      'selected window.'
     exit 1
   fi
 
-  # Find the highest ancestor of the target window below
-  topmost=$(echo -n "$treeout" \
-    | head -n $lineno \
+  echo "$topmost"
+} #}
+
+function get_focused_window_id
+{ #{
+  echo "$(xprop -root -notype _NET_ACTIVE_WINDOW \
+    | grep -Eo '0x[[:xdigit:]]+' \
+    | head -n 1)"
+} #}
+
+function get_current_opacity
+{ #{
+  cur="$(xprop -id "$winid" -notype _NET_WM_WINDOW_OPACITY \
+    | sed -E 's/^_NET_WM_WINDOW_OPACITY = ([0-9]*)$|^.*$/\1/')"
+
+  # Unset opacity means fully opaque
+  test -z "$cur" && cur=0xffffffff
+  cur=$((cur * 100 / 0xffffffff))
+
+  echo "$cur"
+} #}
+
+
+
+function get_opacity
+{ #{
+  echo "$(get_current_opacity)"
+  exit 0
+} #}
+
+function delete_opacity
+{ #{
+  xprop -id "$winid" -remove _NET_WM_WINDOW_OPACITY
+  exit 0
+} #}
+
+function reset_opacity  # Reset opacity of all windows
+{ #{
+    xwininfo -root -tree \
     | sed -n 's/^     \(0x[[:xdigit:]]*\).*/\1/p' \
-    | tail -n 1)
-fi
+    | while IFS=$(printf '\n') read wid;
+      do
+        xprop -id "$wid" -remove _NET_WM_WINDOW_OPACITY
+      done
+    exit 0
+} #}
 
-if test -z "$topmost"; then
-  echo 'Failed to find the highest parent window below root of the' \
-    'selected window.'
+function set_opacity
+{ #{
+  if ! echo "$target_opacity" | grep -qE '^[+-]?[[:digit:]]+%?$'
+  then
+    if test -z "$target_opacity"
+    then echo "No opacity specified" >&2
+    else echo "Invalid opacity specified: $target_opacity" >&2
+    fi
+    exit 1
+  fi
+
+  # strip trailing '%' sign, if any
+  target_opacity="${target_opacity%%%}"
+
+  current_opacity=$(get_current_opacity)
+  echo "$target_opacity" | grep -qE '^[+-]' \
+    && target_opacity=$(( current_opacity + target_opacity ))
+
+  test $target_opacity -lt 0   && target_opacity=0
+  test $target_opacity -gt 100 && target_opacity=100
+
+  target_opacity=$((target_opacity * 0xffffffff / 100))
```

Comment:
We should be consistent with spaces after before the braces inside arithmetic expansions (i.e. `$((expression))` or `$(( expression ))`. Either don't add them for consistency with command substitutions, or add them to differentiate them from command substitutions.

Reply:
I prefer adding spaces. i.e. `$(( expression ))`.

Fixed!